app/src/main/java/org/p2p/wallet/settings/ui/newreset/pin/NewResetPinPresenter.kt

```diff
@@ -0,0 +1,103 @@
+import org.p2p.wallet.common.mvp.BasePresenter
+import org.p2p.wallet.utils.emptyString
+import timber.log.Timber
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+private const val VIBRATE_DURATION = 500L
+private const val PIN_ANIMATION_DURATION = 500L
+
+class NewResetPinPresenter(
+    private val authInteractor: AuthInteractor,
+    private val authLogoutInteractor: AuthLogoutInteractor
+) : BasePresenter<NewResetPinContract.View>(), NewResetPinContract.Presenter {
+
+    private var isCurrentPinConfirmed = false
+    private var createdPin = emptyString()
+
+    override fun setPinCode(pinCode: String) {
+        if (!isCurrentPinConfirmed) {
+            verifyPin(pinCode)
+        } else {
+            resetPin(pinCode)
+        }
+    }
+
+    override fun logout() {
+        launch {
+            authLogoutInteractor.onUserLogout()
+            // Onboarding fragment will figure out if device is saved or not
+            // and do the navigation
+            view?.navigateToOnboarding()
+        }
+    }
+
+    private fun verifyPin(pinCode: String) {
+        launch {
+            try {
+                onSignInResult(authInteractor.signInByPinCode(pinCode))
+            } catch (e: Throwable) {
+                Timber.e(e, "error checking current pin")
+                view?.showIncorrectPinError()
+                view?.vibrate(VIBRATE_DURATION)
+            }
+        }
+    }
+
+    private fun onSignInResult(result: SignInResult) {
+        when (result) {
+            is SignInResult.Success -> {
+                launch {
+                    isCurrentPinConfirmed = true
+                    view?.showPinCorrect()
+                    delay(PIN_ANIMATION_DURATION)
+                    view?.showPinConfirmation()
+                }
+            }
+            SignInResult.WrongPin -> {
+                view?.showIncorrectPinError()
+                view?.vibrate(VIBRATE_DURATION)
+            }
+        }
+    }
+
+    private fun resetPin(pinCode: String) {
+        if (createdPin.isEmpty()) {
+            createdPin = pinCode
+            view?.showPinConfirmed()
+            return
+        }
+
+        if (createdPin != pinCode) {
+            view?.showIncorrectPinError()
+            view?.showMessage(R.string.settings_item_pin_do_not_match)
+            view?.vibrate(VIBRATE_DURATION)
+            return
+        }
+
+        resetPinActually()
+    }
+
+    private fun resetPinActually() {
+        launch {
+            try {
+                authInteractor.resetPin(createdPin)
+                view?.showMessage(R.string.settings_item_pin_changed)
+                view?.showResetSuccess()
+            } catch (e: Throwable) {
+                Timber.e(e, "error setting new pin")
+                view?.showErrorMessage(e)
+                view?.showPinCorrect()
+                view?.vibrate(VIBRATE_DURATION)
+            } finally {
+                createdPin = emptyString()
```

Comment:
nit: do we want to manipulate strings here instead of CharArray? CharArray won't hold the pin code in the String pool and it's safer

Reply:
I reused the implementation which works perfect since it was developed. 
We can consider moving top CharArray as a tech debt. 

Refactoring it now may cause additional bugs, we don't have time